File: CPP/ByTier/D/2066_D2_Club_of_Young_Aircraft_Builders_hard_version.cpp
// Problem: CF 2066 D2 - Club of Young Aircraft Builders (hard version)
// https://codeforces.com/contest/2066/problem/D2

/*
D2. Club of Young Aircraft Builders (hard version)
Algorithm: Dynamic Programming with Combinatorics
Time Complexity: O(n * m * c) per test case
Space Complexity: O(n * m) per test case

This problem involves determining the number of valid ways to fill in missing floor numbers 
(0s) in an array of airplane launch records such that:
1. Each floor's perspective is respected (if c airplanes have been launched up to that point, no more from that floor).
2. The total number of airplanes launched equals m.
3. At any time, for every floor, at least c airplanes must have been launched by that time.

The approach uses dynamic programming where:
- f[i][j] represents the number of ways to fill the first i floors with exactly j airplanes.
- Preprocessing is used for combinations (C[n][k]) and cumulative counts of airplanes per floor.
- For each position, we calculate valid transitions based on constraints from the current and previous floors.

*/

#include <algorithm>
#include <stdio.h>
#define rep(i, a, b) for (int i = (a); i <= (b); i++)
#define per(i, a, b) for (int i = (a); i >= (b); i--)
using namespace std;
const int N = 105, M = 10005, P = 1e9 + 7;
int n, c, m, a[M], C[N][N], f[N][M], ct[N][M], R[N];

void Inc(int &x, int y) {
  if ((x += y) >= P)
    x -= P;
}

void solve() {
  scanf("%d%d%d", &n, &c, &m), C[0][0] = 1;
  rep(i, 1, c) {
    C[i][0] = 1;
    rep(j, 1, c) C[i][j] = (C[i - 1][j - 1] + C[i - 1][j]) % P;
  }
  rep(i, 1, m) scanf("%d", &a[i]);
  rep(i, 1, n) R[i] = 0;
  rep(i, 1, m) R[a[i]] = max(R[a[i]], i); // R[i] stores latest index where floor i launched an airplane
  rep(i, 0, n + 1) rep(j, 0, m + 1) ct[i][j] = 0;
  rep(i, 1, m) ct[a[i]][i]++;
  rep(i, 1, n) rep(j, 1, m) ct[i][j] += ct[i][j - 1]; // prefix sum over columns
  rep(j, 1, m) per(i, n - 1, 1) ct[i][j] += ct[i + 1][j]; // suffix sum over rows
  rep(i, 0, n) rep(j, 0, m) f[i][j] = 0;
  f[0][0] = 1; // Base case: 0 floors, 0 airplanes
  rep(i, 1, n) rep(j, 0, m - c) if (f[i - 1][j]) {
    int r = j + c;
    if (R[i] > r)
      continue; // Not valid due to constraint violation
    int x = ct[i][r] - ct[i + 1][r], y = r - ct[i + 1][r] - j; // Calculating range of possible values
    rep(k, x, y) Inc(f[i][j + k], 1ll * f[i - 1][j] * C[y - x][k - x] % P); // DP transition with binomial coefficients
  }
  printf("%d\n", f[n][m]);
}

int main() {
  int T;
  scanf("%d", &T);
  while (T--)
    solve();
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/2066/2066_D2_Club_of_Young_Aircraft_Builders_hard_version.py
# Problem: CF 2066 D2 - Club of Young Aircraft Builders (hard version)
# https://codeforces.com/contest/2066/problem/D2

"""
D2. Club of Young Aircraft Builders (hard version)

This problem involves counting the number of valid ways to fill in missing airplane launch data,
where each launch is attributed to a specific floor of a building. The constraints are:
- There are n floors, c minimum required airplanes, and m total launched airplanes.
- For each launch at floor i, all residents from floor 1 to i can see it.
- At most one person per floor can launch an airplane before seeing c total airplanes.
- The final arrangement must ensure that every floor sees c or more airplanes by the end.

Approach:
- Dynamic Programming (DP) with state [floor][seen_airplanes]
- Precompute combination numbers C(n, k) for binomial coefficients
- Process each test case using DP over the sequence of launches
- For each step, compute valid transitions where the number of airplanes launched
  up to a given floor meets the constraint that no floor can launch more than c before 
  seeing at least c airplanes from lower floors.

Time Complexity: O(n^3 * m) in worst case per test case due to nested loops.
Space Complexity: O(n^2 + m) for combination table and DP arrays.

Algorithms/Techniques:
- Dynamic Programming
- Precomputed Binomial Coefficients
"""

def solve():
    import sys

    input_data = sys.stdin.read().split()
    if not input_data:
        return
    mod = 10**9 + 7
    # Precompute binomial coefficients C(n, k)
    c = [[0] * 101 for _ in range(101)]
    c[0][0] = 1
    for i in range(1, 101):
        c[i][0] = 1
        for j in range(1, i + 1):
            c[i][j] = (c[i - 1][j - 1] + c[i - 1][j]) % mod
    ptr = 0
    t = int(input_data[ptr])
    ptr += 1
    out_lines = []
    for _ in range(t):
        n = int(input_data[ptr])
        ptr += 1
        C_val = int(input_data[ptr])
        ptr += 1
        m_val = int(input_data[ptr])
        ptr += 1
        a = [0] * (m_val + 1)
        cnt = [[0] * (m_val + 1) for _ in range(n + 1)]
        # Count occurrences of each floor up to position j
        for i in range(1, m_val + 1):
            a[i] = int(input_data[ptr])
            ptr += 1
            for j in range(0, n + 1):
                cnt[j][i] = cnt[j][i - 1]
            cnt[a[i]][i] += 1
        f = [[0] * (m_val + 1) for _ in range(n + 1)]
        f[0][0] = 1
        # DP transition
        for i in range(0, n):
            for j in range(0, m_val + 1):
                if f[i][j] == 0:
                    continue
                r = j + C_val
                if r > m_val:
                    r = m_val
                fix = cnt[i + 1][r]
                whi = r - j
                # Check if valid transition based on floor constraints
                if cnt[i + 1][r] != cnt[i + 1][m_val]:
                    continue
                for k in range(i + 2, n + 1):
                    whi -= cnt[k][r]
                # Compute combinations for filling gaps
                for t_val in range(fix, whi + 1):
                    if t_val > C_val:
                        break
                    if i != n - 1 or t_val == C_val:
                        f[i + 1][j + t_val] = (
                            f[i + 1][j + t_val] + f[i][j] * c[whi - fix][t_val - fix]
                        ) % mod
        out_lines.append(str(f[n][m_val] % mod))
    sys.stdout.write("\n".join(out_lines))


if __name__ == "__main__":
    solve()


# https://github.com/VaHiX/codeForces/